pkg/controller/elasticsearch/driver/pvc.go

```diff
@@ -0,0 +1,122 @@
+package driver
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"github.com/elastic/cloud-on-k8s/pkg/apis/elasticsearch/v1alpha1"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/label"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/sset"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/stringsutil"
+)
+
+// GarbageCollectPVCs ensures PersistentVolumeClaims created for the given es resource are deleted
+// when no longer used, since this is not done automatically by the StatefulSet controller.
+// Related issue in the k8s repo: https://github.com/kubernetes/kubernetes/issues/55045
+// It sets an owner reference to automatically delete PVCs on es deletion, and garbage collects
+// unused PVCs of existing StatefulSets.
+// Note we do **not** delete the corresponding PersistentVolumes but just the PersistentVolumeClaims.
+// PV deletion is left to the responsibility of the storage class reclaim policy.
+func GarbageCollectPVCs(
+	k8sClient k8s.Client,
+	scheme *runtime.Scheme,
+	es v1alpha1.Elasticsearch,
+	actualStatefulSets sset.StatefulSetList,
+	expectedStatefulSets sset.StatefulSetList,
+) error {
+	// PVCs are using the same labels as their corresponding StatefulSet, so we can filter on ES cluster name.
+	var pvcs corev1.PersistentVolumeClaimList
+	if err := k8sClient.List(&client.ListOptions{
+		Namespace:     es.Namespace,
+		LabelSelector: label.NewLabelSelectorForElasticsearch(es),
+	}, &pvcs); err != nil {
+		return err
+	}
+	if err := reconcilePVCsOwnerRef(k8sClient, scheme, es, pvcs.Items); err != nil {
+		return err
+	}
+	return deleteUnusedPVCs(k8sClient, pvcs.Items, actualStatefulSets, expectedStatefulSets)
+}
+
+// reconcilePVCsOwnerRef ensures PVCs created for this Elasticsearch cluster have an owner ref set to
+// the Elasticsearch resource, so they are deleted automatically upon Elasticsearch deletion.
```

Comment:
As it was noted already, in some scenarios ES deletion is not a part of the lifecycle, ie. cluster could be scaled up/down, but never deleted. In such a case, which I think is not uncommon, this change won't help.

Reply:
Not sure I understand your comment; I think this is covered already?

Reply:
I wanted to say that cleaning up after ES resource is deleted might not help in many cases, but yes, it's already covered :)